int32 link1
int32 link2
int32 link3
int32 z